### PR TITLE
[ICV][XLink] - Port changes

### PR DESCRIPTION
**Pull request by DariaMityagina**

### CODE REVIEW

**Review by Maxim-Doronin — APPROVED** [review]
LGTM now

File: inference-engine/thirdparty/movidius/XLink/shared/src/XLinkDispatcher.c [hunk 1]
```diff
@@ -293,6 +322,8 @@ XLinkError_t DispatcherStart(xLinkDeviceHandle_t *deviceHandle)
         mvLog(MVLOG_ERROR,"pthread_attr_destroy error");
     }
 
+    while(((XLink_sem_wait(&schedulerState[idx].eventSchedulerStartSem)) == -1) && errno == EINTR)
+        continue;
```
Comment (Maxim-Doronin) [inline inference-engine/thirdparty/movidius/XLink/shared/src/XLinkDispatcher.c:326]:
```c
    int rc;
    while(((rc = XLink_sem_wait(&schedulerState[idx].eventSchedulerStartSem)) == -1) && errno == EINTR)
        continue;
    if (rc) {
```

File: inference-engine/thirdparty/movidius/XLink/shared/src/XLinkPrivateFields.c [hunk 2]
```diff
@@ -55,7 +56,8 @@ streamDesc_t* getStreamById(void* fd, streamId_t id)
     int stream;
     for (stream = 0; stream < XLINK_MAX_STREAMS; stream++) {
         if (link->availableStreams[stream].id == id) {
-            XLink_sem_wait(&link->availableStreams[stream].sem);
+            while(((XLink_sem_wait(&link->availableStreams[stream].sem)) == -1) && errno == EINTR)
```
Comment (Maxim-Doronin) [inline inference-engine/thirdparty/movidius/XLink/shared/src/XLinkPrivateFields.c:59]: The same. And return NULL

File: inference-engine/thirdparty/movidius/XLink/shared/src/XLinkPrivateFields.c [hunk 3]
```diff
@@ -70,7 +72,8 @@ streamDesc_t* getStreamByName(xLinkDesc_t* link, const char* name)
     for (stream = 0; stream < XLINK_MAX_STREAMS; stream++) {
         if (link->availableStreams[stream].id != INVALID_STREAM_ID &&
             strcmp(link->availableStreams[stream].name, name) == 0) {
-            XLink_sem_wait(&link->availableStreams[stream].sem);
+            while(((XLink_sem_wait(&link->availableStreams[stream].sem)) == -1) && errno == EINTR)
```
Comment (Maxim-Doronin) [inline inference-engine/thirdparty/movidius/XLink/shared/src/XLinkPrivateFields.c:75]: same

**Review by Maxim-Doronin — APPROVED** [review]
minor comments

File: inference-engine/thirdparty/movidius/XLink/shared/src/XLinkDispatcher.c [hunk 4]
```diff
@@ -322,9 +322,12 @@ XLinkError_t DispatcherStart(xLinkDeviceHandle_t *deviceHandle)
         mvLog(MVLOG_ERROR,"pthread_attr_destroy error");
     }
 
-    while(((XLink_sem_wait(&schedulerState[idx].eventSchedulerStartSem)) == -1) && errno == EINTR)
+    int rc = 0;
+    while(((rc = XLink_sem_wait(&schedulerState[idx].eventSchedulerStartSem)) == -1) && errno == EINTR)
         continue;
-    sem_post(&addSchedulerSem);
+    if (rc) {
+        sem_post(&addSchedulerSem);
```
Comment (Maxim-Doronin) [inline inference-engine/thirdparty/movidius/XLink/shared/src/XLinkDispatcher.c:329]: I guess it is a mistake

Comment (DariaMityagina) [reply inference-engine/thirdparty/movidius/XLink/shared/src/XLinkDispatcher.c]: My bad. Fixed.

**Comment by Maxim-Doronin** [comment]
@evgenytalanin-intel @azhogov 
looks like azure builds have hanged. Is it possible to restart them w/o restarting jenkins ci (it is green finally)?

File: inference-engine/thirdparty/movidius/XLink/shared/src/XLinkDispatcher.c [hunk 5]
```diff
@@ -361,19 +393,41 @@ int DispatcherWaitEventComplete(xLinkDeviceHandle_t *deviceHandle)
...
+                Sleep(1);
+#else
+                usleep(1000);
+#endif
+            }
+        }
+    } else {
+        while(((rc = XLink_sem_wait(id)) == -1) && errno == EINTR)
+            continue;
+    }
```
Comment (themarpe) [inline inference-engine/thirdparty/movidius/XLink/shared/src/XLinkDispatcher.c:415]: Hi @Maxim-Doronin @DariaMityagina 
Is this change Windows related? If so, the `sem_timedwait` calls this ported `clock_gettime` function 
```
int clock_gettime(int dummy, struct timespec *spec)
{
    dummy;	// for unreferenced formal parameter warning
    __int64 wintime; GetSystemTimeAsFileTime((FILETIME*)&wintime);
    wintime -= 116444736000000000LL;  //1jan1601 to 1jan1970
    spec->tv_sec = wintime / 10000000LL;           //seconds
    spec->tv_nsec = wintime % 10000000LL * 100;      //nano-seconds
    return 0;
}
```

Which seems to retrieve system clock instead of monotonic (also `win_time.h` defines `CLOCK_MONOTONIC` as 0, so if a check is added, that should be changed as well)

File: inference-engine/thirdparty/movidius/XLink/shared/include/XLink.h [hunk 6]
```diff
@@ -180,6 +188,33 @@ XLinkError_t XLinkReleaseData(streamId_t streamId);
...
+ * @brief Sends a package to initiate the writing of data to a remote stream with timeout in ms
+ * XLinkWriteDataWithTimeout is not fully supported yet. The XLinkWriteData method is called instead.
+ * @warning Actual size of the written data is ALIGN_UP(size, 64)
+ * @param[in] streamId – stream link Id obtained from XLinkOpenStream call
+ * @param[in] buffer – data buffer to be transmitted
+ * @param[in] size – size of the data to be transmitted
+ * @param[in] timeoutMs – timeout for a write operation in milliseconds
+ * @return Status code of the operation: X_LINK_SUCCESS (0) for success
+ */
+XLinkError_t XLinkWriteDataWithTimeout(streamId_t streamId, const uint8_t* buffer, int size, unsigned int timeoutMs);
```
Comment (themarpe) [inline inference-engine/thirdparty/movidius/XLink/shared/include/XLink.h:216]: Looking forward to this being able to be used to gracefully exit a thread that is blocked on this operation, waiting for an incoming/outgoing packet.

Also from my naive testing of adding an equivalent function (by adding just an event with timeout), I had bunch of issues - like events would be left in dispatcher. 

Would this upcoming changes allow for "polling" of packets with specific timeout in a tight loop without any difference compared to a blocking call?
